### ISSUE TITLE
sError in .dsy2dpo(.M2sym(from)) :

### ISSUE THREAD

**Issue by Ashduc** [kdzimm/PseudoreplicationPaper#6]
It is very helpful to have a detailed study like yours that addresses a number of critical questions about differential expression for scRNA-Seq experiments. 
I am running MAST with RE for a categorical response variable where I have 12 donors from condition_1 and 11 donors from condition_2.

When I run MAST for 20 different cell types separately, some of them fail giving the below error,
I wonder if you have encountered such situation and would be glad if you could share any suggestions. Thank you

```
sError in .dsy2dpo(.M2sym(from)) :
  not a positive definite matrix (and positive semidefiniteness is not checked)
In addition: Warning messages:
1: In vcov.merMod(object@fitD) :
  variance-covariance matrix computed from finite-difference Hessian is
not positive definite or contains NA values: falling back to var-cov estimated from RX
2: In vcov.merMod(object@fitD) :
  variance-covariance matrix computed from finite-difference Hessian is
not positive definite or contains NA values: falling back to var-cov estimated from RX
3: In vcov.merMod(object@fitD) :
  variance-covariance matrix computed from finite-difference Hessian is
not positive definite or contains NA values: falling back to var-cov estimated from RX
4: In vcov.merMod(object@fitD) :
  variance-covariance matrix computed from finite-difference Hessian is
not positive definite or contains NA values: falling back to var-cov estimated from RX
5: In vcov.merMod(object@fitD) :
  variance-covariance matrix computed from finite-difference Hessian is
not positive definite or contains NA values: falling back to var-cov estimated from RX
6: In checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv,  :
  Model failed to converge with max|grad| = 0.00204015 (tol = 0.002, component 1)
7: In checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv,  :
  unable to evaluate scaled gradient
8: In checkConv(attr(opt, "derivs"), opt$par, ctrl = control$checkConv,  :
  Problem with Hessian check (infinite or missing values?)
9: In vcov.merMod(object@fitC) :
  Computed variance-covariance matrix problem: not a positive definite matrix (and positive semidefiniteness is not checked);
returning NA matrix
```

**Comment by kdzimm** [comment]
@Ashduc - it is possible that you have either complete separation (all zero in one group, all expressed in the other) or you have genes included that are all zero. Are you filtering out lowly expressed genes? When you get this error does it time out or does it just print the warning and still provides an output?

**Comment by kdzimm** [comment]
Another thing to check is balance across donors for each cell type (it may be the case that you have some donors with a very low number of cells)

**Comment by Ashduc** [comment]
@kdzimm Thank you for the quick response. I am filtering 10% low expression genes. However, when I increase the cutoff to 20% it works. But I think 20% is too stringent. Also to notice, it works fine when Fixed-effect is used for donor_ID. Once I get the error it does not provide any output.

**Comment by kdzimm** [comment]
Stick with a fixed-effect then. Are you finding that you are losing critical genes when you filter 20%? In my experience, within a particular cell type, there are only several hundred genes worth computing tests with.

**Comment by Ashduc** [comment]
@kdzimm 
Thanks
When I use fixed effects, I see some genes that are specific to donor-donor variation become top hits. Which is not our primary focus. (We visually inspected the expressions for individual donors) 
Which basically replicates the output from Seurat's FindAllMarkers() with the wilcox test. 
Also, to note I am inputting log2(raw_counts+1) as the input.

**Comment by kdzimm** [comment]
Using DonorID as a fixed effect should not be replicating the results of FindAllMarkers() with the wilcox test. Your p-values will be many order or magnitudes smaller with FindAllMarkers(), but I guess the rank could be somewhat stable. The donor-to-donor variation may possibly still be there, but those top hits should be there because they are expressed more frequently/abundantly in one condition vs the other, so they may still be worth looking at.

**Comment by Ashduc** [comment]
Yes, the top hits are there. I will look into it more. Thanks for your input.
